### PR TITLE
More intuitive Details block with summary and innerBlocks content

### PR DESCRIPTION
**Pull request by richtabor**
Simplifies the experimental Details block to have the `<summary>` as part of the block itself and Innerblocks for the content, suggested by @talldan in https://github.com/WordPress/gutenberg/pull/45055#issuecomment-1478983941. This is the same method used in the Quote block as well. I'd feel comfortable moving this out of the experiment following this PR. 

### This method makes the Details block simpler by: 
- **Reducing the complexities around style conflicts, as there is one block with design tooling—rather than three.** Noted in https://github.com/WordPress/gutenberg/pull/45055#issuecomment-1468863344). 
- **Allowing the Details block itself to be easily selectable.** In trunk, you have to first select the child "Details Summary" or "Details Content" blocks, then select the parent from the toolbar. 
- **Removing an extraneous tree level of blocks**. as the parent block _is_ the Innerblocks. 
- Negating the need for block locking (and therefore additional icons), as the summary will always be at the top of the Details block.

## What?
1. Removes the Details Summary and Details Content blocks.
2. Moves InnerBlocks capability to the Details block.
3. Tidies up the initial placeholder state and block example.
4. Improves the details icon to better support hi-res devices.
5. Simplifies the block description.

## Testing Instructions

1. Enable the Details block experiment within Gutenberg admin settings.
2. Open a Post or Page.
3. Insert a Details block. 
3. Add a summary.
4. Add blocks to the Details block content.
5. Leverage design tooling on the Details block.
6. See changes on the front-end. 

#### Example, icon, and description: 
CleanShot 2023-04-13 at 13 43 59

#### Empty:
CleanShot 2023-04-13 at 13 38 45

#### Completed: 
CleanShot 2023-04-13 at 13 40 52

#### Completed with additional styling (to show how it can still be flexed using this method): 
CleanShot 2023-04-13 at 13 42 44
CleanShot 2023-04-13 at 13 49 12
CleanShot 2023-04-13 at 13 58 53
CleanShot 2023-04-13 at 14 00 00

### CODE REVIEW

**Comment by richtabor** [comment]
One thing I wanted to add (if anyone has ideas how) is to make `Enter` or `Return` from the `<summary>` RichText component carry you to the innerBlocks.

File: packages/block-library/src/details/style.scss
Comment (talldan) [inline packages/block-library/src/details/style.scss:1]: It looks like the `cursor: pointer;` style for the summary element has been lost from the styles.

**Review by talldan — COMMENTED** [review]
I like the simplicity this brings.

The main loss in styling capability seems to be around styling the summary (border/padding etc.).

A few blocks have similar challenges where they have distinct elements that can't be styled (e.g. the button and input in the search block), so it might be nice to solve at some point.

**Comment by fabiankaegy** [comment]
Yeah I do also like the simplicity of this approach. 

However I worry the loss of the design options of the summary element are going to really limit the styling options you can create for the block. At least without having to resort to custom css.

**Comment by jasmussen** [comment]
> A few blocks have similar challenges where they have distinct elements that can't be styled (e.g. the button and input in the search block), so it might be nice to solve at some point.

This is a solid point, and it's likely to be relevant for various form controls as well. A comments block is never going to be useful without a "Post Comments" button, and the gained simplicity from a flatter structure is commonly going to be useful for these cases. 

One idea for how to solve this is in global styles, surfacing the "Summary" element in a theoretical "Elements" panel, a la #47369. CC: @SaxonF

**Comment by talldan** [comment]
> One idea for how to solve this is in global styles, surfacing the "Summary" element in a theoretical "Elements" panel, a la https://github.com/WordPress/gutenberg/issues/47369.

I understand most of the elements being at the top-level, as they can be shared between blocks, but I wonder if Summary would make sense. I think it's only ever going to be used in the details block.

**Comment by jasmussen** [comment]
Yeah it's not ideal, but maybe it's not too far a stretch in light of something like #44671.

**Comment by richtabor** [comment]
> However I worry the loss of the design options of the summary element are going to really limit the styling options you can create for the block. At least without having to resort to custom css.

I'd say the design tools we have cover the 90% of styling cases for the most part; just look at the example screenshots above I took of me flexing it a bit. 

I think it's worth the hefty win in usability for a slight loss in that 10%.

**Comment by richtabor** [comment]
> I understand most of the elements being at the top-level, as they can be shared between blocks, but I wonder if Summary would make sense. I think it's only ever going to be used in the details block.

I agree, doesn't seem like it fits on the elements front.

**Comment by richtabor** [comment]
> The main loss in styling capability seems to be around styling the summary (border/padding etc.). 

I've wrapped the inner blocks in a group and styled from there to get some of the more unique designs up above. It's not 100% ideal, but is sensible—and reduces the conflicting "where is this style coming from" issues from the separate Details, Summary and Content blocks in trunk.

File: packages/block-library/src/details/style.scss
Comment (richtabor) [reply packages/block-library/src/details/style.scss:1]: Resolved, thanks! I made sure it only does pointer on front-end, as the editor is a RichText.

**Comment by carolinan** [comment]
I am not sure this is more intuitive (Please note I am not in any way a UX expert!) and would prefer to see some user testing results. But I am also unsure of how we could test both versions to see which is preferred.

**Comment by carolinan** [comment]
( And before a call for testing can go out I would prefer to see this issue solved: https://github.com/WordPress/gutenberg/issues/49679)

**Comment by paaljoachim** [comment]
First of all I am testing 
https://github.com/WordPress/gutenberg/pull/45055
through gutenberg.run 
I went to the Gutenberg menu and turned on the Details block experiment and then went back to a page.

-----

Looking at:

> Reducing the complexities around style conflicts, as there is one block with design tooling—rather than three. Noted in Add Details/summary block #45055 (comment)).
> 
> For example, Buttons block does not allow coloring, but the individual Button blocks do. Perhaps we should do the same here. Otherwise you have color and dimensions applied throughout where it's hard to tell what is coming from where.

I believe this should be addressed more in general, as an UX issue that we need to address likely for all of the Block Editor.

> Perhaps the child blocks should not have margin and padding, if the parent does. But block spacing would be nice/relevant on the parent block, and content block perhaps.

I do not understand the logic here. As what you say here would also have relevance for the Columns block with the individual child columns. As for the Columns block it would be nice to be able to fully style the parent and child columns.

> Those styles should probably be applied to a parent Group block, which would nearly prevent—but greatly reduce—the same sort of styling clipping issues. 

There is also a draft PR by @aaronrobertshaw which mentions column child clipping here:
https://github.com/WordPress/gutenberg/pull/40925 

Clipping is an issue, but clipping is a minor issue in relation to giving the user the opportunity to style the parent and child the way they need to. Clipping in general is also something that needs to be addressed down the road. Right now I believe that Aarons draft PR has not been merged because of the clipping issue. Again it is minor compared to allowing the user to style things in the way they need.

Screenshot 2023-04-16 at 10 53 55

---

> Allowing the Details block itself to be easily selectable. In trunk, you have to first select the child "Details Summary" or "Details Content" blocks, then select the parent from the toolbar.

To me this is setup the same way as selecting the Gallery block. There is no click the Details block and have the parent first selected and then click the inner blocks, so this is a general issue for blocks that have inner blocks. That all blocks with inner blocks the first initial click should be the parent. Click again to select the inner/child blocks. 

Styling in importance as I see it when going through this right now.
To be able to style the top parent the Details block itself. - Check that can be done. -
To be able to style the various inner workings of the block. - Check that can be done. -
Thank you.

> Removing an extraneous tree level of blocks. as the parent block is the Innerblocks.

It works for me. As it gives the user the opportunity to style different levels of the Details block. There are so many different use cases to how someone would like to use the Details block. The only conflict I see is the clippings and that is not a big issue.

> Negating the need for block locking (and therefore additional icons), as the summary will always be at the top of the Details block.

This I do not understand. Details the top parent will always be the top.

Btw the following could be an example for an artsie site using the Details block. Giving the user a lot of ways to style as they choose.
Screenshot 2023-04-16 at 11 08 35


Btw

> ......and reduces the conflicting "where is this style coming from" issues from the separate Details, Summary and Content blocks in trunk.

This is more of a general issue that design will need to look closer at. Because there will be various blocks that one will be able to style in many different ways as well as levels inner and parent. Example from the Columns block to where the parent and inner columns have been styled.

Screenshot 2023-04-16 at 11 22 22

Instead of limiting the expression of a user it is better to help give hints as to what the user is styling.

Let me know if there is anything else I should share my thoughts on.

**Comment by aaronrobertshaw** [comment]
> Clipping is an issue, but clipping is a minor issue in relation to giving the user the opportunity to style the parent and child the way they need to... Again it is minor compared to allowing the user to style things in the way they need.

The issue of clipping when changing the shape of a block might seem minor at first glance however with the current inserter being in-canvas, clipping content can cause substantial UX issues as the inserter button can be hidden. I'd argue not being able to easily insert additional blocks is a greater problem than adding a border radius to the container.

I agree we should be able to provide border-radius on more blocks while also better handling the clipping and potential UX impacts there. Joen outlines all this in a more elegant and detailed manner over on the relevant PR (https://github.com/WordPress/gutenberg/pull/40925#issuecomment-1508100067).

Relative to this PR, I don't have any strong opinion as to whether border-radius should be included in the adopted block supports or not. You can see the block inserter being clipped in the following demo.

<details>
<summary>Demo video of inserter being clipped</summary>


https://user-images.githubusercontent.com/60436221/232350587-b365c6c5-da21-4bb4-8ad0-e893ec1a04a2.mp4


</details>

**Comment by carolinan** [comment]
So this clipping happens no matter which version of the block that is used? 🤔 But not if the border-radius is on a wrapping group block?

File: packages/block-library/src/details/save.js
Comment (carolinan) [inline packages/block-library/src/details/save.js]: We can't use translations in save.js

https://github.com/WordPress/gutenberg/pull/45055/files/731a8e802bc2b0c7ba64c8fc1ef0bdaf91a0eee9#r1144295804

**Comment by jasmussen** [comment]
I would also echo: fewer blocks, fewer things to have to click and manage. As far as user testing, I would rather test a baseline that omits the separate summary block, and add it back if during the outreach program people say: "I want to style the summary element separately". It's a clear ux win, the only thing that's unclear for me is whether the ability to style the summary element is valuable or not.

**Comment by richtabor** [comment]
> It's a clear ux win, the only thing that's unclear for me is whether the ability to style the summary element is valuable or not.

It's still possible, just not as intuitive as a standalone block.

**Comment by carolinan** [comment]
It is not so clear that if you do not add a text to "Write summary....." the word "Summary" will show in its place. 

Other than that I believe this can be merged (as experimental) after the border radius block support is removed?

**Comment by richtabor** [comment]
> It is not so clear that if you do not add a text to "Write summary....." the word "Summary" will show in its place.

Yea, its more of a fallback; though I don't think I prefer adding the text in as the default. I'm not sure we add default text for any other block.

**Comment by carolinan** [comment]
When the `<details>` HTML element does not include a summary, the default text it adds is Details not Summary.

Default texts are added for the navigation links  (next /previous)

**Comment by cbirdsong** [comment]
I'm excited this is finally happening! Some stray thoughts about it:

- I think styling the summary element is valuable. There are so many different contexts where you might use a collapse like this, and you might want one that contains a video transcript to look very different than one containing technical details about a project. Having it be a separate block is also useful for adding different appearance options using block styles.

- I find it unintuitive and clunky that you can apply separate styles to both the Details Content block and the Details wrapper block. I'm not sure what the use case is for being able to do this, for instance: 
CleanShot 2023-04-20 at 09 05 40
CleanShot 2023-04-20 at 09 09 46

- The Details Content block should probably support block spacing.

**Comment by richtabor** [comment]
@carolinan I've updated the fallback to use "Details" instead. Mind doing a quick pass again?

**Comment by justintadlock** [comment]
I wasn't aware of this separate PR until today after I was running some design tests with with the current experimental block (which I think is pretty great right now from a design perspective). This was research for a [potential Dev Blog tutorial](https://github.com/WordPress/developer-blog-content/discussions/104).

Styling the Details Summary was **literally** the first thing I did. Styling the Details Content was the second thing I did.

From a design perspective, having individual blocks will provide a ton of flexibility to both theme authors and users. At the very least, I'd think the `<summary>` element needs its own set of design tools. The Details Content less so because the nested blocks would likely cover most tools needed.

With a single, unified block, themers will need to rely more heavily on CSS than on built-in design tools. Things are still wide open with custom block styles and patterns, but users will be limited to whatever the theme offers and fewer tools from core.

**Comment by richtabor** [comment]
> From a design perspective, having individual blocks will provide a ton of flexibility to both theme authors and users.

@justintadlock Are you thinking the design controls in this proposal aren't sufficient? I shared a few instances of how they can be designed at the top of this PR.

**Comment by richtabor** [comment]
> From a design perspective, having individual blocks will provide a ton of flexibility to both theme authors and users. 

But that comes at the cost of experience. 

While it may afford a little more flexibility in design to split into entirely separate blocks, the experience of adding a details block is substantially more complicated _for most people_:

1. It quickly becomes a "where's this style coming from" hodgepodge of block selection and design overrides, which is unhelpful—as styles come from the parent and each level of nested blocks (summary, content, then content innerBlocks). 

2. Block selection is also easier, as in trunk it is difficult to select the parent Details block to move, or style it. You have to select _the space between_ the Summary and Content child blocks.  

The proposal here is along the line of how the Quote block works, where the `cite` is not a separate block, but part of the parent block. I think it's the right move, in favor of a much better experience—but still quite capable Details block.

**Comment by justintadlock** [comment]
> @justintadlock Are you thinking the design controls in this proposal aren't sufficient? I shared a few instances of how they can be designed at the top of this PR.

Correct. At the very least, not being able to style the `<summary>` element separately would be insufficient. I could, of course, do this as a designer with CSS. But if I wanted to style it via the design tools as a user or as a creator with little code knowledge, there doesn't seem to be a way to do it. And, of course, there's the lack of `theme.json` support.

Something like this basic design becomes unnecessarily complicated:

image

I don't think the Details Content block is necessary. You should be able to drop a Group in there, for example, and use it to do whatever is needed.

**Comment by draganescu** [comment]
Should `<summary>` be styled via the elements API, like links, `<cite>` and `<caption>`? I think wanting to visually style the summary separately is a need that comes from wanting to use a group of Details blocks to emulate an accordion, what do you think @justintadlock ?

Instead should we continue the work to add an accordion block, and prioritise here the base usecase where you need only sporadic, isolated details?

Adding the `<summary>` element to the elements API also has the advantage that it's promoting a holistic design tool for a specific element (same as with say links). The downside is that a user wanting to style the summary will have to read the docs and learn where they need to go and do a global edit.

**Comment by carolinan** [comment]
cite and caption are used by more than block.  I think the previous discussions also brought up that the `<summary> `was not suitable for this since it is only used here.

**Comment by justintadlock** [comment]
> Should `<summary>` be styled via the elements API, like links, `<cite>` and `<caption>`? 

So far, what we have are mostly inline types of changes for elements (e.g., colors, typography). `<summary>` would also need things like dimension and border support. Are those things supported at the element level and in the UI? If so, it'd be better than nothing at all, but it wouldn't be ideal.

Would a theme be able to have multiple Details/Summary patterns with different designs (without resorting to custom CSS)? I could, for example, change link colors via block attributes, but how would I change the padding or border for `<summary>` on a pattern-by-pattern basis?

At the global level, it'd be really weird to style it as a user since it's disconnected from its one and only use case. Also, it wouldn't be customizable per block.

> I think wanting to visually style the summary separately is a need that comes from wanting to use a group of Details blocks to emulate an accordion, what do you think @justintadlock ?

I'd want to style the summary for other use cases too. The "accordion" idea was just something I was exploring and not really relevant  to the need to style the element.

> Instead should we continue the work to add an accordion block, and prioritise here the base usecase where you need only sporadic, isolated details?

I personally think an accordion block would fall more under plugin territory. Nevertheless, I'm mostly wanting to make sure the DX and UX is flexible enough for Details/Summary in its base use cases.

**Comment by cbirdsong** [comment]
> I don't think the Details Content block is necessary. You should be able to drop a Group in there, for example, and use it to do whatever is needed.

I think working to reduce the amount of situations where this is required would generally be good, and I definitely wouldn't want to move in the other direction.

I can see how it seems unnecessary since the details content block requires basically no functionality. Could it be implemented as a variant of the group block?

**Comment by carolinan** [comment]
This discussion continues to go in circles after several years, can we **settle** and let the **users** test it? Please? :)

**Comment by fabiankaegy** [comment]
Agreed. So which experiment do we want to run by users, and how do we evaluate whether it is working? 

Do we wait for feedback on the current iteration, which includes the separate summary block, or do we merge this and get feedback without the summary element.

I personally am in favor of testing it with the summary block, but I am not the one making a call on that :)

**Comment by richtabor** [comment]
> or do we merge this and get feedback without the summary element.

My understanding is that it’s not possible to build a semantic and accessible accordion with the details and summary. So I’d be in favor of a Details block that does not lead folks to customizing things in a way that looks like an accordion (this proposal). 

Based on feedback from this proposal we could consider follow-up for more design tooling, instead of starting with more and retracting.

**Comment by carolinan** [comment]
I don't understand what that means, that it shouldn't look like an accordion. 
Opening one details element doesn't close others. That doesn't mean it would be wrong of a user to place three details blocks that has blue backgrounds next to each other. That is not going to cause accessibility issues.

**Comment by jasmussen** [comment]
[Fabian summed it up quite well](https://github.com/WordPress/gutenberg/pull/49808#issuecomment-1531062287): our first goal is to get a version of this block into core, out of experimental, and then get feedback.

For me, that is a very strong argument for keeping this initial version as simple as possible. It is easy to add features and complexity to the block, as it becomes clearer that it is needed. The code for a separate summary block even exists to be resurfaced. We can land this PR without it, get feedback from the plugin, add it back two weeks later if it comes up. The reverse, however, having the separate block and removing it will be that much harder: ensuring deprecations are in place, and even discerning that usability feedback is related to the known challenge of having to select the right layer, vs. something else. This one is hard to beat in terms of simplicity:

Screenshot 2023-05-03 at 09 43 11

The accessibility feedback is another argument to start simple. The Accordion block (#21584) has been stalled for a while due to the complexity of making it properly accessible. On that thread, it was discussed using the details/summary combo as an alternative, but [it was recommended we avoid this](https://github.com/WordPress/gutenberg/issues/21584#issuecomment-660195507). It's a tricky balance: we provide warnings against insufficient contrast, we don't prevent you from choosing bad colors. Similarly we can't prevent people from using the details/summary combo as an accordion replacement. 

But for all that discussion, and because an actual accordion button could still land, and because it is so easy to land a separate summary element later on, I would actually start with less, not more.

**Comment by carolinan** [comment]
Thank you.

And do we consider the border radius clipping to be a bigger issue than this block, [or does the border radius need to be removed?](https://github.com/WordPress/gutenberg/pull/49808/files#diff-bbeb18009d508630b37d7929fd7333c6b80dcebc1481fe045f3836b3b21596fbR31)

**Comment by jasmussen** [comment]
I want to note one thing in reviewing this which we'll want to fix. It maybe my env, but I can't press space inside the summary element:

Screenshot 2023-05-03 at 09 43 20

**Comment by carolinan** [comment]
Yes there is an open issue for that, it only happens with Firefox. https://github.com/WordPress/gutenberg/issues/49679

**Comment by jasmussen** [comment]
> And do we consider the border radius clipping to be a bigger issue than this block, [or does the border radius need to be removed?](https://github.com/WordPress/gutenberg/pull/49808/files#diff-bbeb18009d508630b37d7929fd7333c6b80dcebc1481fe045f3836b3b21596fbR31)

If the radius adds complexity, and to the same argument as starting simple, I'd have no problem removing that for now and exploring adding it back separately, with whatever challenges that might entail.

**Comment by jasmussen** [comment]
> Yes there is an open issue for that, it only happens with Firefox. https://github.com/WordPress/gutenberg/issues/49679

Oh interesting, I experience this in Chrome 112.

**Comment by carolinan** [comment]
Ah, yes I see the problem.
[Both the summary tag name and the prevent default is added directly to the richText](https://github.com/WordPress/gutenberg/pull/49808/files#diff-20d83620db8197c679e5617a73e78de6306074dff49588bb00fd5ee35a17793bR62)

**Comment by richtabor** [comment]
> Ah, yes I see the problem.
> [Both the summary tag name and the prevent default is added directly to the richText](https://github.com/WordPress/gutenberg/pull/49808/files#diff-20d83620db8197c679e5617a73e78de6306074dff49588bb00fd5ee35a17793bR62)

Ah, is that why you didn’t use the summary tagName with RichText originally?

**Comment by carolinan** [comment]
Yes, but, there are issues with that solution too.

**Comment by mtias** [comment]
Not sure how helpful this feedback is at this stage, but I think we should lean on simplicity as much as possible, not on customization. There's value in getting the ability to have content that con be toggle in the hands of users. It should look like regular content, like a paragraph that can be toggled, not a full fledged accordion that has other implications. More customization would be misleading in that case because people use it to make the collapsed section look like a heading, different from the rest of the content, which would be problematic. Can we reduce it to its most simplest and go from there?

**Comment by richtabor** [comment]
> I want to note one thing in reviewing this which we'll want to fix. It maybe my env, but I can't press space inside the summary element:

I reverted back to wrapping the RichText in a summary element (the original method), which resolves this.

**Comment by richtabor** [comment]
> And do we consider the border radius clipping to be a bigger issue than this block, [or does the border radius need to be removed?](https://github.com/WordPress/gutenberg/pull/49808/files#diff-bbeb18009d508630b37d7929fd7333c6b80dcebc1481fe045f3836b3b21596fbR31)

Also removed border radius support.

**Comment by richtabor** [comment]
I think it's ready for another good look.

**Review by talldan — COMMENTED** [review]
Code-wise this looks good, but I wonder whether a deprecation should be added now that the details block has been in a plugin release or two.

It is behind an experiment toggle, but it seems to break in a fairly bad way that doesn't allow users to easily fix their content.

What do you think?

**Review by talldan — APPROVED** [review]
I'll approve it and let you make the call.

**Comment by richtabor** [comment]
> Code-wise this looks good, but I wonder whether a deprecation should be added now that the details block has been in a plugin release or two. What do you think?

I don't think we need the deprecation.

**Comment by paaljoachim** [comment]
What about using lessons learnt in regards to the Details block and continue with a Tabs block?
As it is a nice block to also get into Gutenberg.
https://github.com/WordPress/gutenberg/issues/34079

**Pull request merged**
